README.md
# NewtonROMReader

This archive contains the PCB layout and firmware source code for my Apple 
Newton MessagePad ROM board reader/programmer.

The Newton ROM Reader reads the contents of Apple MessagePad MP2x00 ROM
boards. Besides the well known US and German ROM, this programmer can read 
prototype ROM boards for further research.

Using the alternative Flash ROM Board, the ROM Reader can also program the
Flash chips and create new ROM Boards with patched or alternative 
operating systems. 

## CAUTION

 - Use an original Arduino Due. Do *NOT* use a Mega 2560!
 - Never insert or remove the ROM card while the programmer is powered!
 - All operations are final! There is no undo! There is no confirmation.

## Reader/Programmer

Assembling the Programmer requires some advanced soldering skills. The pin 
headers must be soldered around the outside of the board facing down. They 
must be flush with the PCB on the top, or the ROM board will not fit.

The SD Card connector is soldered to the underside of the board. The top
of the connector should be insulated using an ample amount of tape, avoiding
a short circuit with header pins on the Arduino. ALternatively, the colliding
pins on the Arduino board can be cut off.

The DIMM memory socket is more difficult to solder. The pin pitch is relatively
small. To make sure that the soldering is correct, the first tool to run
in the Programmer should be the short circuit test.

The last solder part is easy, but must not be forgotten. It's a jumper that 
sets the power to 3.2V. Connect the two eyes marked with an asterisk (*).

```
:
:   Lower left corner of the PCB
|
| ooo ooo
| ooo **o
|     ...... ........
+--------------------- - -
``` 

The Programmer board should be mounted on an Arduino Due. It *MUST NOT* be 
used with an Arduino Mega 1280/2560. The Mega runs at 5V and will likely
destroy any ROM, Flash, or SD card inserted into the Programmer. The Due runs
at the required 3.2V. If the large USB-A port of the Arduino blocks access to 
the SD Card slot, you have the wrong board!

The Programmer communicates with the user via USB connection to the Programming
port. No extra power connection is required. There is no extra software for
the Programmer. Users must use a serial terminal program to connect via the 
Programming USB port (57600 bps/baud 8N1).

## Firmware Source Code

The source code comes as an Arduino sketch `.ino` file and a number of 
additional source and header files. The "SdFat" library by Bill Greimann must be 
installed within the Arduino system `Tools > Manage Libraries...`.

Load the `.ino` file and set the board to `Arduino Due (Programming Port)`, 
then choose the right port from the `Tools > Port` menu.

Open the serial monitor `Tools > Serial Monitor` and press `Ctrl-U` to upload
the firmware to the Programmer.  

## PCB Layout

The Programmer PCB was designed with Eagle CAD 7.7.0 Stadard Edition. It's a
two layer board with no exeptional requirements.

The SO-DIMM 72 modules are not very common anymore and are somewhat hard to 
find. The SD Card socket is a standard SMD part. Pin headers should be 
square shaped to fit the Arduino receptacle well.

## Usage

The Programmer writes a 90's style menu over the USB serial port connection.
Functions can be selected by typing the command key and Return. 

 - `h`: Check all connectors on the board for short circuits. This tool helps
        verifying that there are no bridges in the SO DIMM connector soldering.
        *Never run this command with a ROM board or SD Card inserted!*
 - `0`: Select the ROM area from 0x00000000 to 0x00800000 which corresponds
        to the regular ROM and Apple system ROM extension (REx).
 - `1`: This page can be programmed with an extension block of up to 8MB.
 - `2`, `3`: Experimental blocks. These should be mapped to 0x10000000 and 
        can contain 8MB extensions each. This however is not yet verified.
 - `p`: Print the contents of the Flash as a hex dump with ASCII. Pressing 
        `return` will print the next block. Pressing any other key followed
        by `enter` will bring you back to the main menu. 
 - `c`: Check if the Flash memory is empty (all addresses return 0xFFFFFFFF).
 - `e`: Erase all sectors in the selected Flash area.
 - `w`: Write the content of a file on SD card or on an USB Drive to the 
        current Flash page. Filenames for pages 0 to 3 are set to "rom.bin",
        "rex1.bin", "rex2.bin", and "rex3.bin".
 - 'v': Verify the Flash content with the corresponding file on SD Card or 
        USB Drive.
  
## ToDo's

 - the SO DIMM footprint is really hard to solder, the pads are too big and 
   solder tends to short the pins
 - the LED can not be controlled because it is connected to a ROM Board output
 - address and data pins should be connected to 32 bit registers, making the 
   pin-by-pin addressing obsolete and the programmer much faster
 - the SPI port should be connected to a hardware SPI
 - there should be an option for a display and four buttons
 - the SD Card connecter must have pull-ups on most pins, and the Card 
   Detect and Write Protect pins should be connected to some pin
 


CHANGES.md
## Versions

### Version 0.3 

Version 0.3 was mostly rewritten, adding support for USB Drives that can be 
connected to the native USB port via OTG cable. If no SD Card is found, we
check for a connected USB Drive. USB and SD drives can be hot swapped.

Menus were restructured, and menu commands were changed (sorry about that).
There is no longer a confirmation required for any operation, making it 
very easy to erase or overwrite a ROM with one keypress.

Pins are now set to inputs with a pullup after every operation, making an
accidental removal of a ROM card without powering down less harmful 
for the card.

### Version 0.2 

Version 0.2 is obsolete.

### Version 0.1 

The initial version is 0.1 which can erase, write, and verify the first ROM 
block and supports SD Cards.


ROMCard.cpp


#include "ROMCard.h"

int gCurrentPage = 0;
uint32_t gCurrentAddress = 0;

static bool sAddressPinsUndefined = true;

std::vector <Page> gPageList = {
  { 0x00000000, 0x00800000, "ROM page", "rom.bin" },
  { 0x00800000, 0x01000000, "ROM Extension 1", "rex1.bin" },
  { 0x02000000, 0x02800000, "ROM Extension 2", "rex2.bin" },
  { 0x02800000, 0x03000000, "ROM Extension 3", "rex3.bin" },
  { 0x00000000, 0x01000000, "ROM & ext1", "romrex1.bin" },
  { 0x02000000, 0x03000000, "ext2 & ext3", "rex23.bin" },
//{ 0x00000000, 0x02000000, "full card", "card.bin" },
};


void dataBus(bool active)
{
  int mode = active ? OUTPUT : INPUT_PULLUP;
//  int mode = active ? OUTPUT : INPUT;
  pinMode(eggD0, mode);
  pinMode(eggD1, mode);
  pinMode(eggD2, mode);
  pinMode(eggD3, mode);
  pinMode(eggD4, mode);
  pinMode(eggD5, mode);
  pinMode(eggD6, mode);
  pinMode(eggD7, mode);
  pinMode(eggD8, mode);
  pinMode(eggD9, mode);
  pinMode(eggD10, mode);
  pinMode(eggD11, mode);
  pinMode(eggD12, mode);
  pinMode(eggD13, mode);
  pinMode(eggD14, mode);
  pinMode(eggD15, mode);
  pinMode(eggD16, mode);
  pinMode(eggD17, mode);
  pinMode(eggD18, mode);
  pinMode(eggD19, mode);
  pinMode(eggD20, mode);
  pinMode(eggD21, mode);
  pinMode(eggD22, mode);
  pinMode(eggD23, mode);
  pinMode(eggD24, mode);
  pinMode(eggD25, mode);
  pinMode(eggD26, mode);
  pinMode(eggD27, mode);
  pinMode(eggD28, mode);
  pinMode(eggD29, mode);
  pinMode(eggD30, mode);
  pinMode(eggD31, mode);
}

void activateDataBusWrite() { dataBus(true); }
void activateDataBusRead() { dataBus(false); }
void deactivateDataBus() { dataBus(false); }

void addressBus(bool active)
{
  int mode = active ? OUTPUT : INPUT_PULLUP;
  pinMode(eggA2, mode);    // 0x00000004
  pinMode(eggA3, mode);    // 0x00000008
  pinMode(eggA4, mode);    // 0x00000010
  pinMode(eggA5, mode);    // 0x00000020
  pinMode(eggA6, mode);    // 0x00000040
  pinMode(eggA7, mode);    // 0x00000080
  pinMode(eggA8, mode);    // 0x00000100
  pinMode(eggA9, mode);    // 0x00000200
  pinMode(eggA10, mode);   // 0x00000400
  pinMode(eggA11, mode);   // 0x00000800
  pinMode(eggA12, mode);   // 0x00001000
  pinMode(eggA13, mode);   // 0x00002000
  pinMode(eggA14, mode);   // 0x00004000
  pinMode(eggA15, mode);   // 0x00008000
  pinMode(eggA16, mode);   // 0x00010000
  pinMode(eggA17, mode);   // 0x00020000
  pinMode(eggA18, mode);   // 0x00040000
  pinMode(eggA19, mode);   // 0x00080000
  pinMode(eggA20, mode);   // 0x00100000
  pinMode(eggA21, mode);   // 0x00200000
  pinMode(eggA22, mode);   // 0x00400000
  pinMode(eggA23, mode);   // 0x00800000
  pinMode(eggA24, mode);   // 0x01000000
  sAddressPinsUndefined = true;
}

void activateAddressBus() { addressBus(true); }
void deactivateAddressBus() { addressBus(false); }

/**
 * Set an address on the address bus.
 * 
 * \param[in] inAddress valid addresses are 0-32MB ind 4 byte steps
 */
void setAddress(uint32_t inAddress)
{
#if 1 // slower version
  uint32_t a = inAddress;
  a >>= 1;
  a >>= 1;
  digitalWrite(eggA2, a & 1); a >>= 1;
  digitalWrite(eggA3, a & 1); a >>= 1;
  digitalWrite(eggA4, a & 1); a >>= 1;
  digitalWrite(eggA5, a & 1); a >>= 1;
  digitalWrite(eggA6, a & 1); a >>= 1;
  digitalWrite(eggA7, a & 1); a >>= 1;
  digitalWrite(eggA8, a & 1); a >>= 1;
  digitalWrite(eggA9, a & 1); a >>= 1;
  digitalWrite(eggA10, a & 1); a >>= 1;
  digitalWrite(eggA11, a & 1); a >>= 1;
  digitalWrite(eggA12, a & 1); a >>= 1;
  digitalWrite(eggA13, a & 1); a >>= 1;
  digitalWrite(eggA14, a & 1); a >>= 1;
  digitalWrite(eggA15, a & 1); a >>= 1;
  digitalWrite(eggA16, a & 1); a >>= 1;
  digitalWrite(eggA17, a & 1); a >>= 1;
  digitalWrite(eggA18, a & 1); a >>= 1;
  digitalWrite(eggA19, a & 1); a >>= 1;
  digitalWrite(eggA20, a & 1); a >>= 1;
  digitalWrite(eggA21, a & 1); a >>= 1;
  digitalWrite(eggA22, a & 1); a >>= 1;
  digitalWrite(eggA23, a & 1); a >>= 1;
  digitalWrite(eggA24, a & 1); a >>= 1;
#elif 0 // faster version
  if (sAddressPinsUndefined)
    gCurrentAddress = ~inAddress;
  uint32_t a = inAddress;
  uint32_t m = inAddress ^ gCurrentAddress;
  a >>= 1; m >>= 1;
  a >>= 1; m >>= 1;
  if (m&1) digitalWrite(eggA2, a & 1); a >>= 1; m >>= 1;
  if (m&1) digitalWrite(eggA3, a & 1); a >>= 1; m >>= 1;
  if (m&1) digitalWrite(eggA4, a & 1); a >>= 1; m >>= 1;
  if (m&1) digitalWrite(eggA5, a & 1); a >>= 1; m >>= 1;
  if (m&1) digitalWrite(eggA6, a & 1); a >>= 1; m >>= 1;
  if (m&1) digitalWrite(eggA7, a & 1); a >>= 1; m >>= 1;
  if (m&1) digitalWrite(eggA8, a & 1); a >>= 1; m >>= 1;
  if (m&1) digitalWrite(eggA9, a & 1); a >>= 1; m >>= 1;
  if (m&1) digitalWrite(eggA10, a & 1); a >>= 1; m >>= 1;
  if (m&1) digitalWrite(eggA11, a & 1); a >>= 1; m >>= 1;
  if (m&1) digitalWrite(eggA12, a & 1); a >>= 1; m >>= 1;
  if (m&1) digitalWrite(eggA13, a & 1); a >>= 1; m >>= 1;
  if (m&1) digitalWrite(eggA14, a & 1); a >>= 1; m >>= 1;
  if (m&1) digitalWrite(eggA15, a & 1); a >>= 1; m >>= 1;
  if (m&1) digitalWrite(eggA16, a & 1); a >>= 1; m >>= 1;
  if (m&1) digitalWrite(eggA17, a & 1); a >>= 1; m >>= 1;
  if (m&1) digitalWrite(eggA18, a & 1); a >>= 1; m >>= 1;
  if (m&1) digitalWrite(eggA19, a & 1); a >>= 1; m >>= 1;
  if (m&1) digitalWrite(eggA20, a & 1); a >>= 1; m >>= 1;
  if (m&1) digitalWrite(eggA21, a & 1); a >>= 1; m >>= 1;
  if (m&1) digitalWrite(eggA22, a & 1); a >>= 1; m >>= 1;
  if (m&1) digitalWrite(eggA23, a & 1); a >>= 1; m >>= 1;
  if (m&1) digitalWrite(eggA24, a & 1); a >>= 1; m >>= 1;
#else
  // make sure that all address pins are refreshed
  if (sAddressPinsUndefined)
    gCurrentAddress = ~inAddress;

  // update only the pins that changed
  uint32_t a = inAddress;
  uint32_t m = inAddress ^ gCurrentAddress;
  a >>= 1; m >>= 1;
  a >>= 1; m >>= 1;
  for (int i=kAddrMin; i<=kAddrMax; i++) {
    if (m&1) digitalWrite(gAdrressLut[i], a & 1); 
    a >>= 1; m >>= 1;
  }
#endif
  // set the global address, so we can optimize pin update
  // also, gCurrentAddress is used to determine the corresponding CS
  gCurrentAddress = inAddress;
}

void activateControlBus()
{
  digitalWrite(eggRESET, 1);
  pinMode(eggRESET, OUTPUT); // inv
  digitalWrite(eggROM_CS_0, 1);
  pinMode(eggROM_CS_0, OUTPUT); // inv
  digitalWrite(eggROM_CS_1, 1);
  pinMode(eggROM_CS_1, OUTPUT); // inv
  digitalWrite(eggROM_IO_RD, 1);
  pinMode(eggROM_IO_RD, OUTPUT);
  digitalWrite(eggROM_IO_WR, 1);
  pinMode(eggROM_IO_WR, OUTPUT);
  pinMode(eggROM_IO_INT, INPUT);
  pinMode(eggROM_IO_RDY, INPUT);
  digitalWrite(eggPOWER_ENABLE, 1);
  pinMode(eggPOWER_ENABLE, OUTPUT);
}

void deactivateControlBus()
{
  pinMode(eggRESET, INPUT_PULLUP);
  pinMode(eggROM_CS_0, INPUT_PULLUP);
  pinMode(eggROM_CS_1, INPUT_PULLUP);
  pinMode(eggROM_IO_RD, INPUT_PULLUP);
  pinMode(eggROM_IO_WR, INPUT_PULLUP);
  pinMode(eggROM_IO_INT, INPUT_PULLUP);
  pinMode(eggROM_IO_RDY, INPUT_PULLUP);
  pinMode(eggPOWER_ENABLE, INPUT_PULLUP);
}

void deactivateROMBus()
{
  deactivateDataBus();
  deactivateAddressBus();
  deactivateControlBus();
}

void setupROMBus()
{
  deactivateROMBus();
}

uint32_t readWord(uint32_t inAddr)
{
  setAddress(inAddr);
  return readWord();
}

uint32_t readWord()
{
  register uint32_t w = 0;
  
  if (gCurrentAddress & kAddrToCSMask)
    digitalWrite(eggROM_CS_1, 0); // select the upper half of the Flash chips
  else
    digitalWrite(eggROM_CS_0, 0); // select the lower half of the Flash chips
    
#if 0 // universal
  digitalWrite(eggROM_IO_RD, 0); // read operation
  w = w << 1; if (digitalRead(eggD31)) w += 1;
  w = w << 1; if (digitalRead(eggD30)) w += 1;
  w = w << 1; if (digitalRead(eggD29)) w += 1;
  w = w << 1; if (digitalRead(eggD28)) w += 1;
  w = w << 1; if (digitalRead(eggD27)) w += 1;
  w = w << 1; if (digitalRead(eggD26)) w += 1;
  w = w << 1; if (digitalRead(eggD25)) w += 1;
  w = w << 1; if (digitalRead(eggD24)) w += 1;
  w = w << 1; if (digitalRead(eggD23)) w += 1;
  w = w << 1; if (digitalRead(eggD22)) w += 1;
  w = w << 1; if (digitalRead(eggD21)) w += 1;
  w = w << 1; if (digitalRead(eggD20)) w += 1;
  w = w << 1; if (digitalRead(eggD19)) w += 1;
  w = w << 1; if (digitalRead(eggD18)) w += 1;
  w = w << 1; if (digitalRead(eggD17)) w += 1;
  w = w << 1; if (digitalRead(eggD16)) w += 1;
  w = w << 1; if (digitalRead(eggD15)) w += 1;
  w = w << 1; if (digitalRead(eggD14)) w += 1;
  w = w << 1; if (digitalRead(eggD13)) w += 1;
  w = w << 1; if (digitalRead(eggD12)) w += 1;
  w = w << 1; if (digitalRead(eggD11)) w += 1;
  w = w << 1; if (digitalRead(eggD10)) w += 1;
  w = w << 1; if (digitalRead(eggD9)) w += 1;
  w = w << 1; if (digitalRead(eggD8)) w += 1;
  w = w << 1; if (digitalRead(eggD7)) w += 1;
  w = w << 1; if (digitalRead(eggD6)) w += 1;
  w = w << 1; if (digitalRead(eggD5)) w += 1;
  w = w << 1; if (digitalRead(eggD4)) w += 1;
  w = w << 1; if (digitalRead(eggD3)) w += 1;
  w = w << 1; if (digitalRead(eggD2)) w += 1;
  w = w << 1; if (digitalRead(eggD1)) w += 1;
  w = w << 1; if (digitalRead(eggD0)) w += 1;
  digitalWrite(eggROM_IO_RD, 1); // end of read operation
#else // Due specific (saving about 40% of time on "Verify")
  PIOB->PIO_CODR = PIO_CODR_P15; // digitalWrite(eggROM_IO_RD, 0); // read operation
  asm(" nop\n nop\n nop\n"); // I don't know why I need a delay here, but it will read faulty data without it
  asm(" nop\n nop\n nop\n");
  uint32_t portA = PIOA->PIO_PDSR;
  uint32_t portB = PIOB->PIO_PDSR;
  uint32_t portC = PIOC->PIO_PDSR;
  uint32_t portD = PIOD->PIO_PDSR;
  PIOB->PIO_SODR = PIO_SODR_P15; // digitalWrite(eggROM_IO_RD, 1); // PB15, end of read operation 
  if (portA & PIO_PDSR_P0)  w |= PIO_PDSR_P18; // const int eggD18 = 69;  // PA0   A15;
  if (portA & PIO_PDSR_P2)  w |= PIO_PDSR_P10; // const int eggD10 = A7;  // PA2
  if (portA & PIO_PDSR_P4)  w |= PIO_PDSR_P9;  // const int eggD9  = A5;  // PA4
  if (portA & PIO_PDSR_P6)  w |= PIO_PDSR_P8;  // const int eggD8  = A4;  // PA6
  if (portA & PIO_PDSR_P7)  w |= PIO_PDSR_P23; // const int eggD23 = 31;  // PA7
  if (portA & PIO_PDSR_P10) w |= PIO_PDSR_P13; // const int eggD13 = 19;  // PA10
  if (portA & PIO_PDSR_P12) w |= PIO_PDSR_P12; // const int eggD12 = 17;  // PA12
  if (portA & PIO_PDSR_P14) w |= PIO_PDSR_P19; // const int eggD19 = 23;  // PA14
  if (portA & PIO_PDSR_P20) w |= PIO_PDSR_P31; // const int eggD31 = 43;  // PA20
  if (portA & PIO_PDSR_P22) w |= PIO_PDSR_P7;  // const int eggD7  = A3;  // PA22
  if (portA & PIO_PDSR_P24) w |= PIO_PDSR_P6;  // const int eggD6  = A1;  // PA24

  if (portB & PIO_PDSR_P13) w |= PIO_PDSR_P14; // const int eggD14 = 21;  // PB13
  if (portB & PIO_PDSR_P16) w |= PIO_PDSR_P17; // const int eggD17 = 67;  // PB16  A13;
  if (portB & PIO_PDSR_P18) w |= PIO_PDSR_P15; // const int eggD15 = A9;  // PB18
  if (portB & PIO_PDSR_P20) w |= PIO_PDSR_P16; // const int eggD16 = A11; // PB20
  if (portB & PIO_PDSR_P25) w |= PIO_PDSR_P5;  // const int eggD5  = 2;   // PB25

  if (portC & PIO_PDSR_P1)  w |= PIO_PDSR_P24; // const int eggD24 = 33;  // PC1
  if (portC & PIO_PDSR_P3)  w |= PIO_PDSR_P25; // const int eggD25 = 35;  // PC3
  if (portC & PIO_PDSR_P5)  w |= PIO_PDSR_P26; // const int eggD26 = 37;  // PC5
  if (portC & PIO_PDSR_P7)  w |= PIO_PDSR_P27; // const int eggD27 = 39;  // PC7
  if (portC & PIO_PDSR_P14) w |= PIO_PDSR_P28; // const int eggD28 = 49;  // PC14
  if (portC & PIO_PDSR_P16) w |= PIO_PDSR_P29; // const int eggD29 = 47;  // PC16
  if (portC & PIO_PDSR_P18) w |= PIO_PDSR_P30; // const int eggD30 = 45;  // PC18
  if (portC & PIO_PDSR_P22) w |= PIO_PDSR_P2;  // const int eggD2  = 8;   // PC22
  if (portC & PIO_PDSR_P24) w |= PIO_PDSR_P3;  // const int eggD3  = 6;   // PC24
  if (portC & PIO_PDSR_P26) w |= PIO_PDSR_P4;  // const int eggD4  = 4;   // PC26
  if (portC & PIO_PDSR_P29) w |= PIO_PDSR_P1;  // const int eggD1  = 10;  // PC29

  if (portD & PIO_PDSR_P0)  w |= PIO_PDSR_P20; // const int eggD20 = 25;  // PD0
  if (portD & PIO_PDSR_P2)  w |= PIO_PDSR_P21; // const int eggD21 = 27;  // PD2
  if (portD & PIO_PDSR_P5)  w |= PIO_PDSR_P11; // const int eggD11 = 15;  // PD5
  if (portD & PIO_PDSR_P6)  w |= PIO_PDSR_P22; // const int eggD22 = 29;  // PD6
  if (portD & PIO_PDSR_P8)  w |= PIO_PDSR_P0;  // const int eggD0  = 12;  // PD8
#endif
  
  if (gCurrentAddress & kAddrToCSMask)
    digitalWrite(eggROM_CS_1, 1); // deselect the upper half of the Flash chips
  else
    digitalWrite(eggROM_CS_0, 1); // deselect the lower half of the Flash chips
    
  return w;
}

void writeWord(uint32_t a, uint32_t w)
{
  setAddress(a);
  activateDataBusWrite();
  digitalWrite(eggD0, w & 1); w = w >> 1;
  digitalWrite(eggD1, w & 1); w = w >> 1;
  digitalWrite(eggD2, w & 1); w = w >> 1;
  digitalWrite(eggD3, w & 1); w = w >> 1;
  digitalWrite(eggD4, w & 1); w = w >> 1;
  digitalWrite(eggD5, w & 1); w = w >> 1;
  digitalWrite(eggD6, w & 1); w = w >> 1;
  digitalWrite(eggD7, w & 1); w = w >> 1;
  digitalWrite(eggD8, w & 1); w = w >> 1;
  digitalWrite(eggD9, w & 1); w = w >> 1;
  digitalWrite(eggD10, w & 1); w = w >> 1;
  digitalWrite(eggD11, w & 1); w = w >> 1;
  digitalWrite(eggD12, w & 1); w = w >> 1;
  digitalWrite(eggD13, w & 1); w = w >> 1;
  digitalWrite(eggD14, w & 1); w = w >> 1;
  digitalWrite(eggD15, w & 1); w = w >> 1;
  digitalWrite(eggD16, w & 1); w = w >> 1;
  digitalWrite(eggD17, w & 1); w = w >> 1;
  digitalWrite(eggD18, w & 1); w = w >> 1;
  digitalWrite(eggD19, w & 1); w = w >> 1;
  digitalWrite(eggD20, w & 1); w = w >> 1;
  digitalWrite(eggD21, w & 1); w = w >> 1;
  digitalWrite(eggD22, w & 1); w = w >> 1;
  digitalWrite(eggD23, w & 1); w = w >> 1;
  digitalWrite(eggD24, w & 1); w = w >> 1;
  digitalWrite(eggD25, w & 1); w = w >> 1;
  digitalWrite(eggD26, w & 1); w = w >> 1;
  digitalWrite(eggD27, w & 1); w = w >> 1;
  digitalWrite(eggD28, w & 1); w = w >> 1;
  digitalWrite(eggD29, w & 1); w = w >> 1;
  digitalWrite(eggD30, w & 1); w = w >> 1;
  digitalWrite(eggD31, w & 1); w = w >> 1;

  if (gCurrentAddress & kAddrToCSMask)
    digitalWrite(eggROM_CS_1, 0); // select the upper half of the Flash chips
  else
    digitalWrite(eggROM_CS_0, 0); // select the lower half of the Flash chips
    
  digitalWrite(eggROM_IO_WR, 0); // read operation
  digitalWrite(eggROM_IO_WR, 1); // end of read operation
  
  if (gCurrentAddress & kAddrToCSMask)
    digitalWrite(eggROM_CS_1, 1); // deselect the upper half of the Flash chips
  else
    digitalWrite(eggROM_CS_0, 1); // deselect the lower half of the Flash chips
    
  activateDataBusRead();
}
